File: golang/013.go
package main

import "fmt"

func findMissingNum(arr []int)int{
	max, min := arr[0], arr[0]
	missingNum := -1
	for _,ch := range arr{
		if ch > max{
			max = ch
		}
		if ch < min{
			min = ch
		}
	}
	//cover for the fact that indexing starts at 0 and max-min returns one less
	array := make([]int, max-min+2)
	fmt.Println(array)

	for _,ch := range arr {
		array[ch] = 1
	}

	for i:=1; i < len(array); i++{
		if array[i] == 0{
			missingNum = i
		}
	}

	return missingNum
}

func main(){
	fmt.Println(findMissingNum([]int{3, 7, 1, 2, 8, 4, 5}))
}

File: python/013.py
def findMissingNum(arr):
    # change name of variables as max and min are inbuilt in python
    max_num, min_num = max(arr), min(arr)

# Not needed thus
    # for x in arr:
    #     if x > max:
    #         max = x
    #     if x < min:
    #         min = x

    # create a num set, set in python have one elem rep 8,8 comes to 8
    # newArr = [None] * (max-min+1)
    num_set = set(arr)

    for x in range(min_num, max_num+1):
        if x not in num_set:
            return x


    return -1

print(findMissingNum([3, 7, 1, 2, 8, 4, 5]))